src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,169 @@
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        T[] array = (T[])list.toArray();
+        for (int k = 0; k < array.length; k++) {
+            insertLast(array[k]);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return getNode(index).data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> node = getNode(index);
+        T oldValue = node.data;
+        node.data = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return removing(getNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> first = head; first != null; first = first.next) {
+                if (first.data == null) {
+                    removing(first);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> first = head; first != null; first = first.next) {
+                if (object.equals(first.data)) {
+                    removing(first);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public void insertLast(T data) {
+        Node<T> tmp = new Node<T>(tail, data, null);
+        if (tail != null) {
+            tail.next = tmp;
+        }
+        tail = tmp;
+        if (head == null) {
+            head = tmp;
+        }
+        size++;
+    }
+
+    public void insertBefore(T element, Node<T> follow) {
+        Node<T> prior = follow.previous;
+        Node<T> newNode = new Node<>(prior, element, follow);
+        follow.previous = newNode;
+        if (prior == null) {
+            head = newNode;
+        } else {
+            prior.next = newNode;
+        }
+        size++;
+    }
+
+    T removing(Node<T> removed) {
+        final T element = removed.data;
+        Node<T> next = removed.next;
+        Node<T> prev = removed.previous;
+
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+            removed.previous = null;
+        }
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.previous = prev;
+            removed.next = null;
+        }
+        removed.data = null;
+        size--;
+        return element;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> current = this.head;
+        int counter = 0;
+        while (current != null & counter != index) {
```

Comment:
let's iterate our LL from head to middle or from tail to middle depending on the index

Reply:
fixed